File: Cavern.QuickEQ/FFTCache.cs
using System;

using Cavern.QuickEQ.Utilities;
using Cavern.Utilities;

namespace Cavern.QuickEQ {
    /// <summary>Precalculated constants and preallocated recursion arrays for a given FFT size.</summary>
    public class FFTCache : IDisposable {
        /// <summary>Cosines.</summary>
        internal float[] cos;
        /// <summary>Sines.</summary>
        internal float[] sin;
        /// <summary>Preallocated even split array.</summary>
        internal Complex[] even;
        /// <summary>Preallocated odd split array.</summary>
        internal Complex[] odd;
        /// <summary>Preallocated worker array.</summary>
        internal Complex[] temp;

        /// <summary>C++ FFT cache class memory address to be passed to <see cref="CavernQuickEQAmp"/>.</summary>
        internal IntPtr Native { get; private set; } = new IntPtr(0);

        /// <summary>FFT cache constructor.</summary>
        public FFTCache(int size) {
            if (CavernAmp.Available) {
                Native = CavernQuickEQAmp.FFTCache_Create(size);
                return;
            }

            int halfSize = size / 2;
            double step = -2 * Math.PI / size;
            cos = new float[halfSize];
            sin = new float[halfSize];
            for (int i = 0; i < halfSize; ++i) {
                double rotation = i * step;
                cos[i] = (float)Math.Cos(rotation);
                sin[i] = (float)Math.Sin(rotation);
            }
            even = new Complex[halfSize];
            odd = new Complex[halfSize];
            temp = new Complex[size];
        }

        /// <summary>Free all used resources if there is any.</summary>
        public void Dispose() {
            if (Native.ToInt64() != 0)
                CavernQuickEQAmp.FFTCache_Dispose(Native);
        }
    }
}

File: CavernAmp/fftcache.cpp
#include <new>
#include <math.h>
#include <stdint.h>
#include <stdlib.h>

#include "fftcache.h"
#include "main.h"
#include "qmath.h"

FFTCache::FFTCache(const int fftSize) {
    depth = fftSize / 2;
    double step = -2 * M_PI / fftSize;
    sin = new float[depth];
    cos = new float[depth];
    for (int i = 0; i < depth; ++i) {
        float rotation = i * step;
        cos[i] = cosf(rotation);
        sin[i] = sinf(rotation);
    }

    int maxDepth = log2(fftSize);
    even = new Complex*[maxDepth];
    odd = new Complex*[maxDepth];
    for (int idepth = 0; idepth < maxDepth; idepth++) {
        even[idepth] = new Complex[1 << idepth];
        odd[idepth] = new Complex[1 << idepth];
    }
}

int FFTCache::size() const {
    return depth;
}

FFTCache::~FFTCache() {
    delete[] sin;
    delete[] cos;
    for (int idepth = 0, maxDepth = log2(depth * 2); idepth < maxDepth; idepth++) {
        delete[] even[idepth];
        delete[] odd[idepth];
    }
    delete[] even;
    delete[] odd;
}

FFTCache* DLL_EXPORT FFTCache_Create(const int fftSize) {
    return new FFTCache(fftSize);
}

int DLL_EXPORT FFTCache_Size(const FFTCache *cache) {
    return cache->size();
}

void DLL_EXPORT FFTCache_Dispose(FFTCache *cache) {
    delete cache;
}
